File: src/main/java/pl/dmichalski/algorithms/_7_binary_search/BinarySearchService.java
package pl.dmichalski.algorithms._7_binary_search;

class BinarySearchService {

    /**
     * O(log n) time complexity
     */
    Integer search(int[] sortedValues, int numberToSearch) {
        if (sortedValues == null) {
            return -1;
        }

        int min = 0;
        int max = sortedValues.length - 1;

        while (min <= max) {
            int middle = (min + max) / 2;
            int currentElement = sortedValues[middle];

            if (currentElement < numberToSearch) {
                min = middle + 1;
            } else if (currentElement > numberToSearch) {
                max = middle - 1;
            } else {
                return middle;
            }
        }

        return -1;
    }
}

File: src/main/kotlin/pl/dmichalski/algorithms/_7_binary_search/BinarySearchService.kt
package pl.dmichalski.algorithms._7_binary_search

internal class BinarySearchService {

    /**
     * O(log n) complexity
     */
    fun search(sortedValues: IntArray?, numberToSearch: Int): Int {
        if (sortedValues == null) {
            return -1
        }

        var min = 0
        var max = sortedValues.size - 1
        while (min <= max) {
            val middle = (min + max) / 2
            val currentElement = sortedValues[middle]
            when {
                currentElement < numberToSearch -> {
                    min = middle + 1
                }
                currentElement > numberToSearch -> {
                    max = middle - 1
                }
                else -> {
                    return middle
                }
            }
        }

        return -1
    }
}
